### PR TITLE
mimirsbrunn: Load administratives regions from osm by level

### PR DESCRIPTION
**Pull request by azime**

### CODE REVIEW

**Comment by azime** [comment]
TODO:
-> Load coordinates of administratives regions
-> purge Elasticserach
-> Insert administartives regions in Elasticserach

File: src/bin/osm2mimir.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+static USAGE: &'static str = "
+Usage:
+    osm2mimir --input=<pbf-file> --admin=<admin-level>...
+
+Options:
+    -h, --help                               Show this message.
+    -i <pbf-file>, --input=<pbf-file>        OSM PBF file.
+    -a <admin-level>, --admin=<admin-level>  Admin levels to keep.
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: -l, --level seems better.

File: src/bin/osm2mimir.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+            if !is_admin {
+                continue;
+            }
+            let level = relation.tags
+                                .get("admin_level")
+                                .and_then(|s| s.parse().ok())
+                                .unwrap_or(0);
+            if !levels.contains(&level) {
+                continue;
+            }
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: maybe we should just skip with a message when we can't read the level:
```rust
let level = relation.tags
    .get("admin_level")
    .and_then(|s| s.parse().ok());
let level = match level {
    None => {
        println!("invalid admin_level for {}: {:?}", relation.id, relation.tags.get("admin_level"));
        continue;
    }
    Some(l) => l,
};
```

File: src/bin/osm2mimir.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            };
+            let zip_code = match relation.tags.get("addr:postcode") {
+                Some(val) => &val[..],
+                _ => "",
+            };
+
+            let admin = mimirsbrunn::Admin {
+                id: admin_id,
+                level: level,
+                name: relation.tags.get("name").unwrap_or(&"NC".to_string()).to_string(),
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: I think an admin without name is just useless, thus, I think we can do something as for admin_level with a log and continue if we don't have this value.

File: src/bin/osm2mimir.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    }
+    return administrative_regions;
+}
+
+fn main() {
+    println!("importing adminstrative region into Mimir");
+    let args: Args = docopt::Docopt::new(USAGE)
+                         .and_then(|d| d.decode())
+                         .unwrap_or_else(|e| e.exit());
+    let map = args.flag_admin.iter().cloned().collect();
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: that's not a map but a set, and the same is not so explicit. Maybe `levels` is a better name?

File: src/lib.rs [hunk 5]
```diff
@@ -114,25 +129,31 @@ fn bulk_index<'a, T, I>(url: &str, mut iter: I) -> Result<u32, curl::ErrCode>
...
     Ok(nb)
 }
 
 fn upsert<T: Incr>(elt: &T, map: &mut HashMap<String, T>) {
     match map.entry(elt.id().to_string()) {
-        Vacant(e) => { e.insert(elt.clone()); }
-        Occupied(mut e) => e.get_mut().incr()
+        Vacant(e) => {
+            e.insert(elt.clone());
+        }
```
Comment (TeXitoi) [inline src/lib.rs]: I know it's not your code, but the {} are quite useless:
```rust
Vacant(e) => e.insert(elt.clone()),
```

File: src/bin/osm2mimir.rs (hunk 2 above)
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: Even better with the contains:
```rust
let level = relation.tags
    .get("admin_level")
    .and_then(|s| s.parse().ok());
let level = match level {
    None => {
        println!("invalid admin_level for {}: {:?}", relation.id, relation.tags.get("admin_level"));
        continue;
    }
    Some(ref l) if !levels.contains(l) => continue,
    Some(l) => l,
};
```

File: src/bin/osm2mimir.rs (hunk 1 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: Ok

File: src/bin/osm2mimir.rs (hunk 2 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: OK

File: src/bin/osm2mimir.rs (hunk 3 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: OK

File: src/bin/osm2mimir.rs (hunk 4 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: OK

File: src/lib.rs (hunk 5 above)
Comment (azime) [inline src/lib.rs]: OK, 
cargo fmt !!

File: src/bin/osm2mimir.rs [hunk 6]
```diff
@@ -1,77 +1,146 @@
...
+                                .and_then(|s| s.parse().ok());
+			let level = match level {
+			    None => {
+			        info!("invalid admin_level for relation {}: admin_level {:?}", relation.id, relation.tags.get("admin_level"));
+			        continue;
+			    }
+			    Some(l) => l,
+			};
+            // administrative region without levelval
+            //println!("level: {:?}", level);
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: remove

File: src/bin/osm2mimir.rs [hunk 7]
```diff
@@ -1,77 +1,146 @@
...
             if !levels.contains(&level) {
+            	info!("admin_level ignored for relation {}:  admin_level {:?} .", relation.id, level);
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: that's verbose, no?

File: src/bin/osm2mimir.rs [hunk 8]
```diff
@@ -1,77 +1,146 @@
...
                 continue;
             }
+            // administrative region with name ?
+            let name = match relation.tags.get("name") {
+                Some(val) => val,
+                None => {
+                	info!("adminstrative regione without name for relation {}:  admin_level {:?} ignored .", relation.id, level);
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: level in a i32, please print it with `{}` and not `{:?}`

File: src/bin/osm2mimir.rs [hunk 9]
```diff
@@ -1,77 +1,146 @@
...
     }
     return administrative_regions;
 }
 
 fn main() {
-    println!("importing adminstrative region into Mimir");
+	env_logger::init().unwrap();
+	
+	debug!("importing adminstrative region into Mimir");
+    //println!("importing adminstrative region into Mimir");
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: remove

File: src/bin/osm2mimir.rs [hunk 10]
```diff
@@ -1,77 +1,146 @@
...
+			        continue;
+			    }
+			    Some(l) => l,
+			};
+            // administrative region without levelval
+            //println!("level: {:?}", level);
             if !levels.contains(&level) {
+            	info!("admin_level ignored for relation {}:  admin_level {:?} .", relation.id, level);
                 continue;
             }
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: you can put this test inside the previous match as said earlier

File: src/bin/osm2mimir.rs [hunk 11]
```diff
@@ -1,77 +1,146 @@
...
-                                .and_then(|s| s.parse().ok())
-                                .unwrap_or(0);
+                                .and_then(|s| s.parse().ok());
+			let level = match level {
+			    None => {
+			        info!("invalid admin_level for relation {}: admin_level {:?}", relation.id, relation.tags.get("admin_level"));
+			        continue;
+			    }
+			    Some(l) => l,
+			};
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: strange indentation

File: src/bin/osm2mimir.rs [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        OsmObj::Node(ref node) => OsmId::Node(node.id),
+        OsmObj::Way(ref way) => OsmId::Way(way.id),
+        OsmObj::Relation(ref rel) => OsmId::Relation(rel.id),
+    }
+}
+
+fn update_coordinates(filename: &String, admins: &mut AdminsMap) {
+	if admins.is_empty() {
+		return;
+	}
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: strange indent

File: src/bin/osm2mimir.rs (hunk 6 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: OK

File: src/bin/osm2mimir.rs (hunk 8 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: OK

File: src/bin/osm2mimir.rs (hunk 9 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: OK

File: src/bin/osm2mimir.rs (hunk 12 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 11 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs (hunk 10 above)
Comment (azime) [inline src/bin/osm2mimir.rs]: ok

File: src/bin/osm2mimir.rs [hunk 13]
```diff
@@ -78,32 +75,33 @@ fn administartive_regions(filename: &String, levels: &HashSet<u32>) -> AdminsMap
...
+                Some(l) => l,
+            };
             // administrative region with name ?
             let name = match relation.tags.get("name") {
                 Some(val) => val,
                 None => {
-                	info!("adminstrative regione without name for relation {}:  admin_level {:?} ignored .", relation.id, level);
-                	continue;
+                    info!("adminstrative regione without name for relation {}:  admin_level {} \
+                           ignored .",
```
Comment (TeXitoi) [inline src/bin/osm2mimir.rs]: regione -> region, ignored . -> ignored.

**Pull request merged**